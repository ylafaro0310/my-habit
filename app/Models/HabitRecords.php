<?php

namespace App\Models;

use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;

class HabitRecords {
    public $validationRules = [
        'habit_id' => 'required|integer',
        'completed_at' => 'required|integer',
        'is_skipped' => 'required|boolean',
    ];
    
    function __construct(){
        $this->pdo = DB::connection()->getPdo();
        $this->table_name = 'habit_records';
    }

    public function index(){
        $sth = $this->pdo->prepare("select * from $this->table_name");
        $sth->execute();
        $result = $sth->fetchAll();
        Log::debug($result);
        return $result;
    }

    public function store($params){
        $columns = implode(', ',array_keys($params));
        $values = implode(', ',array_map(function($v){return "'$v'";},array_values($params)));
        $sth = $this->pdo->prepare("insert into $this->table_name (${columns}) values (${values})");
        return $sth->execute();
    }

    public function destroy($id){
        $sth = $this->pdo->prepare("delete from $this->table_name where id = ${id}");
        return $sth->execute();
    }
}