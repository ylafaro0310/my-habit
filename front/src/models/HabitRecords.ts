import { List, Record } from 'immutable';
import { Dayjs } from 'dayjs';

import dayjs from '../lib/dayjs-ja';
import { JSObject } from '../types/Common';

export class HabitRecord extends Record<{
  habitId: number;
  completedAt: Dayjs;
  isSkipped: boolean;
}>({
  habitId: 0,
  completedAt: dayjs(),
  isSkipped: false,
}) {
  static fromResponse(response: JSObject): HabitRecord {
    const params = { ...response };
    params['completedAt'] = dayjs(response['completedAt'], 'YYYY-MM-DD HH:mm:ss');
    return new HabitRecord(params);
  }
}

export default class HabitRecords extends Record<{
  items: List<HabitRecord>;
}>({
  items: List(),
}) {
  static fromResponse(response: JSObject): HabitRecords {
    const params = { ...response };
    params.items = List(params.habitRecords.map((item: JSObject) => HabitRecord.fromResponse(item)));
    return new HabitRecords(params);
  }
  getList(): List<HabitRecord> {
    return this.get('items');
  }
  filterById(id: number): HabitRecords {
    return this.set(
      'items',
      this.get('items').filter(habitRecord => habitRecord.habitId === id),
    );
  }
  filterByCompletedAt(date: Dayjs): HabitRecords {
    return this.set(
      'items',
      this.get('items').filter(habitRecord => habitRecord.completedAt.isSame(date, 'd')),
    );
  }
}
