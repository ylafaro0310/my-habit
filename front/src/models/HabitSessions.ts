import { Record } from 'immutable';
import dayjs, { Dayjs } from 'dayjs';
import { JSObject } from '../types/Common';

export default class HabitSessions extends Record<{
    habit_id: number;
    completed_at: Dayjs;
    working_hours: number; // 単位は「分」
}>({
    habit_id: 0,
    completed_at: dayjs(),
    working_hours: 0,
}) {
  static fromResponse(response: JSObject) {
    const params = { ...response };
    return new HabitSessions(params);
  }
}