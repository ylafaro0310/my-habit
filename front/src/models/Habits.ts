import { List, Record } from 'immutable';
import dayjs, { Dayjs } from 'dayjs';
import { JSObject } from '../types/Common';
import HabitRecords from './HabitRecords';

export class Habit extends Record<{
    id: number;
    habit_name: string;
    repeat_type: string;
    repeat_value: number;
    started_date: Dayjs;
    target_time: number; // 単位は「分」
    time_of_day: string;
    consecutive_days: number;
    consecutive_weeks: number | null;
}>({
    id: 0,
    habit_name: '',
    repeat_type: 'day_of_week',
    repeat_value: 127,
    started_date: dayjs(),
    target_time: 0,
    time_of_day: '',
    consecutive_days: 0,
    consecutive_weeks: null,
}) {
  static fromResponse(response: JSObject) {
    const params = { ...response };
    return new Habit(params);
  }
  isCompleted(date: Dayjs,habit_records: HabitRecords): boolean {
    return habit_records.filterById(this.id).filterByCompletedAt(date = dayjs('2020-01-08')).getList().size > 0
  }
}

export default class Habits extends Record<{
  items: List<Habit>;
}>({
  items: List(),
}){
  static fromResponse(response: JSObject) {
    const params = { ...response };
    params.items = List(params.habits.map((item: JSObject) => Habit.fromResponse(item)));
    return new Habits(params);
  }
  getList(): List<Habit>{
    return this.get('items');
  }
}