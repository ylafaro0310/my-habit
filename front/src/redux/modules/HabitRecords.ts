import actionCreatorFactory from 'typescript-fsa';
import { reducerWithInitialState } from 'typescript-fsa-reducers';

import HabitRecords from '../../models/HabitRecords';

// Action Creator
const actionCreator = actionCreatorFactory('HabitRecord');
export const HabitRecordsActions = {
  getHabitRecords: actionCreator<object | undefined>('getHabitRecords'),
  setHabitRecords: actionCreator<HabitRecords>('setHabitRecords'),
  addHabitRecord: actionCreator<object>('addHabitRecord'),
  removeHabitRecord: actionCreator<number>('removeHabitRecord'),
};

// Reducers
export const habitRecordsReducer = reducerWithInitialState(new HabitRecords()).case(
  HabitRecordsActions.setHabitRecords,
  (state, payload) => {
    return state.set('items', payload.getList());
  },
);
