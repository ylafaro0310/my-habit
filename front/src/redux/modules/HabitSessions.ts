import actionCreatorFactory from 'typescript-fsa';
import { reducerWithInitialState } from 'typescript-fsa-reducers';

import HabitSessions from '../../models/HabitSessions';

// Action Creator
const actionCreator = actionCreatorFactory('HabitSession');
export const HabitSessionsActions = {
  getHabitSessions: actionCreator<{ habitId: number; values?: object }>(
    'getHabitSessions',
  ),
  setHabitSessions: actionCreator<HabitSessions>('setHabitSessions'),
  addHabitSession: actionCreator<{ habitId: number; values: object }>(
    'addHabitSession',
  ),
  updateHabitSession: actionCreator<{
    habitId: number;
    habitSessionId: number;
    values: object;
  }>('updateHabitSession'),
  removeHabitSession: actionCreator<{
    habitId: number;
    habitSessionId: number;
  }>('removeHabitSession'),
  formInitialize: actionCreator<{ habitId: number; habitSessionId: number }>(
    'habitSessionFormInitialize',
  ),
};

// Reducers
export const HabitSessionsReducer = reducerWithInitialState(
  new HabitSessions(),
).case(HabitSessionsActions.setHabitSessions, (state, payload) => {
  return state.set('items', payload.getList());
});
