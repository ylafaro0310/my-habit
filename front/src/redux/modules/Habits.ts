import actionCreatorFactory from 'typescript-fsa';
import { reducerWithInitialState } from 'typescript-fsa-reducers';

import Habits from '../../models/Habits';

// Action Creator
const actionCreator = actionCreatorFactory('Habit');
export const HabitsActions = {
  getHabits: actionCreator<void>('getHabits'),
  setHabits: actionCreator<Habits>('setHabits'),
  addHabit: actionCreator<object>('addHabit'),
  updateHabit: actionCreator<{ habitId: number; values: object }>(
    'updateHabit',
  ),
  removeHabit: actionCreator<number>('removeHabit'),
  formInitialize: actionCreator<number>('habitFormInitialize'),
};

// Reducers
export const habitsReducer = reducerWithInitialState(new Habits()).case(
  HabitsActions.setHabits,
  (state, payload) => {
    return state.set('items', payload.getList());
  },
);
