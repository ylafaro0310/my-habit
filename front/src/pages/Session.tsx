import React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { RouteComponentProps } from 'react-router-dom';
import styled from 'styled-components';

import dayjs from '../lib/dayjs-ja';
import { State } from '../redux/store';
import HabitSessions from '../models/HabitSessions';
import { HabitSessionsActions } from '../redux/modules/HabitSessions';

type SessionProps = {
  habitSessions: HabitSessions;
  dispatch: Dispatch;
} & RouteComponentProps<{ habitId: string }>;
export class Session extends React.Component<SessionProps> {
  componentDidMount() {
    const { habitId } = this.props.match.params;
    const { dispatch } = this.props;
    dispatch(
      HabitSessionsActions.getHabitSessions({
        habitId: Number(habitId),
      }),
    );
  }

  render() {
    const { habitSessions } = this.props;
    let sessions;
    if (habitSessions) {
      let keys = habitSessions
        .getList()
        .map(elem => dayjs(elem.completedAt).format('YYYY-MM-DD'));
      keys = keys.filter((v, i, self) => self.indexOf(v) === i);
      keys = keys.sort((a, b) => (dayjs(a).isBefore(dayjs(b)) ? 1 : -1));
      sessions = keys.map((date, key) => (
        <List key={key}>
          <ListLabel>{date}</ListLabel>
          {habitSessions
            .getList()
            .filter(
              elem => dayjs(elem.completedAt).format('YYYY-MM-DD') === date,
            )
            .sort((a, b) =>
              dayjs(a.completedAt).isAfter(dayjs(b.completedAt)) ? 1 : -1,
            )
            .map((elem, key) => (
              <ListItem key={key}>
                <div>{elem.workingMinutes + 'åˆ†'}</div>
                <div>{dayjs(elem.completedAt).format('HH:mm')}</div>
              </ListItem>
            ))}
        </List>
      ));
    }
    return sessions ? sessions : null;
  }
}

export default connect((state: State) => ({
  habitSessions: state.habitSessions,
}))(Session);

const List = styled.div`
  width: 100%;
`;
const ListLabel = styled.div`
  padding: 0.2rem 0.5rem 0.2rem 0.5rem;
  background: #999;
  font-weight: 500;
`;
const ListItem = styled.div`
  padding: 0.2rem 0.5rem 0.2rem 0.5rem;
  display: flex;
  justify-content: space-between;
`;
