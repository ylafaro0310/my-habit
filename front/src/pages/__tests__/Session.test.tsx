import React from 'react';
import { shallow } from 'enzyme';
import MockDate from 'mockdate';

import dayjs from '../../lib/dayjs-ja';
import { Session } from '../Session';
import Habits from '../../models/Habits';
import HabitRecords from '../../models/HabitRecords';
import HabitSessions from '../../models/HabitSessions';

MockDate.set('2020-02-08');

const data = {
  habits: [
    { id: 1, habitName: '読書する', startedAt: dayjs() },
    { id: 2, habitName: '筋トレ', startedAt: dayjs() },
    { id: 3, habitName: '新しいCDを1枚聴く', startedAt: dayjs() },
  ],
  habitSessions: [
    { habitId: 2, workingMinutes: 5, completedAt: dayjs() },
    { habitId: 1, workingMinutes: 10, completedAt: dayjs('2020-01-08') },
  ],
};

const props = {
  habits: Habits.fromResponse(data),
  habitSessions: HabitSessions.fromResponse(data),
  match: {
    params: {
      id: 1,
    },
  },
  dispatch: jest.fn(),
};

describe('Page: Session', () => {
  it('Snapshot test', () => {
    const wrapper = shallow(<Session {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
