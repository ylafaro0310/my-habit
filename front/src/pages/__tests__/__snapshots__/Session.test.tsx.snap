// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page: Session Snapshot test 1`] = `
<Fragment>
  <HabitHeader
    backTo="/records"
    habitName=""
    nextTo="/habits/undefined/sessions"
  />
  <styled.div
    key="0"
  >
    <styled.div>
      2020-02-08
    </styled.div>
    <styled.div
      key="0"
    >
      <Link
        to="/habits/2/sessions/0"
      >
        <styled.div>
          <div>
            5分
          </div>
          <div>
            09:00
          </div>
        </styled.div>
      </Link>
    </styled.div>
  </styled.div>
  <styled.div
    key="1"
  >
    <styled.div>
      2020-01-08
    </styled.div>
    <styled.div
      key="0"
    >
      <Link
        to="/habits/1/sessions/0"
      >
        <styled.div>
          <div>
            10分
          </div>
          <div>
            00:00
          </div>
        </styled.div>
      </Link>
    </styled.div>
  </styled.div>
</Fragment>
`;
